woocommerce/woocommerce-gateway-stripe
# Map WooCommerce states and Payment Request states

Fixes #1477
Fixes #880

This is a fix adapted from 1423-gh-Automattic/woocommerce-payments.

#### Changes proposed in this Pull Request:

**The problem:** Payment Request API states are different than WooCommerce states for some countries. E.g:
`China: Anhui Sheng` and `Anhui / 安徽` in WC. On top of that, WC states are also translated, making it impossible to match in certain scenarios.

This is a temporary solution for matching Payment Request API states and WooCommerce states until we have this list available in core.

I created a class only with a constant array and didn't add its related methods because this is similar to what we're going to get from core when the list is available in a future release of WC.

Right now, to get the current list of translated states, you can do `WC()->countries->get_states( $country )`. In the future, we should be able to do something like `WC()->countries->get_payment_request_states( $country )` and use that instead of this class in the normalization functions.

This PR also introduces a more descriptive error message in case the state field is expected, but not supported by the Payment Request API during checkout. This can happen for many countries and the results are different depending on which browser you're using the Payment Request API.

There can be some cases where there are shipping zones configured for unsupported countries, and in that case, no shipping methods will be available. - This is part of another fix that will be handled in the shipping settings screen, in a different PR.

#### Testing instructions

- Make sure you have "Payment Request Button" enabled in Payments > Settings.
- Make sure test mode is enabled, so you can make test transactions.
- Go to an HTTPS product URL and click the "Buy now" button.
- Add a shipping or billing address in China, select any state and click "Pay" - notice the order goes through with the correct shipping state
- You can test with other previously unsupported countries: `EG` (Egypt), `HK` (Hong Kong), `IR` (Iran).
- Safari doesn't provide a dropdown list of states, so if you try typing the state code or state name without accents, it should work where it would previously fail. E.g: Country: `BR` (Brazil), State: `Parana` (correct spelling is Paraná). Or `ES` (Spain), State: `Malaga` (correct spelling is Málaga).
- You can also test the error notice for unsupported countries trying to checkout with a shipping address in `AO` (Angola), `BD` (Bangladesh), or `ZA` (South Africa).

-------------------
